File: 0633-sum-of-square-numbers/0633-sum-of-square-numbers.java
class Solution {
    public boolean judgeSquareSum(int c) {
        /*if(c < 3) return true;
        
        long l = 0;
        long r = c-1;
        
        while(l <= r){
            if(l*l + r*r == c){
                return true;
            }
            if(l*l + r*r < c){
                l++;
            }else{
                r--;
            }
        }
        
        return false;*/
        
        //O(n)
        HashSet<Integer> set = new HashSet<Integer>();
        
        for(int i = 0; i <= (int)Math.sqrt(c); i++){
            set.add(c - i*i);
            if(set.contains(i*i))
                return true;            
        }
        return false;
    }
}

File: 0633-sum-of-square-numbers/0633-sum-of-square-numbers.kt

class Solution {
    fun judgeSquareSum(c: Int): Boolean {
        
        // approach 2 - Two pointer
        var l = 0
        var r = Math.sqrt(c.toDouble()).toInt()
        
        while(l <= r){
            val sum = l*l + r*r
            if(sum == c) return true
            
            // added sum < 0 for sum overflow cases
            if(sum > c || sum < 0) r--
            else l++
        }
        return false
        
        
        // approach 1 - Linear O(sqrt of c) - similar to Two Sum
        /*val set = HashSet<Int>()
        
        val root = Math.sqrt(c.toDouble()).toInt();
        for(i in 0..root){
            set.add(c - i*i);
            if(set.contains(i*i)) return true;
        }
        return false;*/
        
    }
}